types/sequencer.go

```diff
@@ -1,23 +1,162 @@
+		s.Sequencers = append(s.Sequencers, proposer)
+	}
+}
+
+func (s *SequencerSet) SetSequencers(bondedSet []*types.Validator) {
+	s.Sequencers = bondedSet
+}
+
+// GetByAddress
+func (s *SequencerSet) GetByAddress(addr []byte) *types.Validator {
+	for _, val := range s.Sequencers {
+		if bytes.Equal(val.Address, addr) {
+			return val
+		}
+	}
+	return nil
+}
+
+// HasAddress returns true if address given is in the validator set, false -
+// otherwise.
+func (s *SequencerSet) HasAddress(address []byte) bool {
+	return s.GetByAddress(address) != nil
+}
+
+// ToValSet returns the sequencer set as a tendermint ValidatorSet.
+// used for PRC client compatibility
+func (s *SequencerSet) ToValSet() *types.ValidatorSet {
+	set := types.NewValidatorSet(s.Sequencers)
+	set.Proposer = s.Proposer
+	return set.CopyIncrementProposerPriority(1)
+}
+
+// ToProto
+func (s *SequencerSet) ToProto() (*pb.SequencerSet, error) {
+	protoSet := new(pb.SequencerSet)
+	seqsProto := make([]*cmtproto.Validator, len(s.Sequencers))
+	for i := 0; i < len(s.Sequencers); i++ {
+		valp, err := s.Sequencers[i].ToProto()
+		if err != nil {
+			return nil, fmt.Errorf("toProto: validatorSet error: %w", err)
+		}
+		seqsProto[i] = valp
+	}
+	protoSet.Sequencers = seqsProto
+
+	if s.Proposer != nil {
+		valProposer, err := s.Proposer.ToProto()
+		if err != nil {
+			return nil, fmt.Errorf("toProto: validatorSet proposer error: %w", err)
+		}
+		protoSet.Proposer = valProposer
+	}
+
+	return protoSet, nil
+}
+
+func (s *SequencerSet) FromProto(protoSet pb.SequencerSet) error {
+	seqs := make([]*types.Validator, len(protoSet.Sequencers))
+	for i, valProto := range protoSet.Sequencers {
+		val, err := types.ValidatorFromProto(valProto)
+		if err != nil {
+			return fmt.Errorf("fromProto: validatorSet error: %w", err)
+		}
+		seqs[i] = val
+	}
+
+	var proposer *types.Validator
+	if protoSet.Proposer != nil {
+		valProposer, err := types.ValidatorFromProto(protoSet.Proposer)
+		if err != nil {
+			return fmt.Errorf("fromProto: validatorSet proposer error: %w", err)
+		}
+		proposer = valProposer
+	}
+
+	s.Sequencers = seqs
+	s.Proposer = proposer
+	s.ProposerHash = GetHash(proposer)
+
+	return nil
+}
+
+func (s *SequencerSet) String() string {
+	return fmt.Sprintf("SequencerSet{Validators: %v, Proposer: %v, ProposerHash: %v}", s.Sequencers, s.Proposer, s.ProposerHash)
+}
+
+func GetHash(seq *types.Validator) []byte {
+	if seq == nil {
+		return []byte{}
+	}
+	// we take a set with only the proposer and hash it
+	tempProposerSet := types.NewValidatorSet([]*types.Validator{seq})
+	return tempProposerSet.Hash()
+}
+
+// Makes a copy of the validator list.
+func (s *SequencerSet) Copy() *SequencerSet {
+	// copy the validators
+	valsCopy := make([]*types.Validator, len(s.Sequencers))
+	for i, val := range s.Sequencers {
+		valsCopy[i] = val.Copy()
+	}
+
+	return &SequencerSet{
+		Sequencers:   valsCopy,
+		Proposer:     s.Proposer,
+		ProposerHash: s.ProposerHash,
+	}
+}
```

Comment:
not really a safe copy since old data is still used?
Can you document purpose of this and deep/shallow